src/services/accounts/AccountsBalanceInfoService.ts

```diff
@@ -25,6 +27,88 @@ export class AccountsBalanceInfoService extends AbstractService {
 	): Promise<IAccountBalanceInfo> {
 		const { api } = this;
 
+		// Retrieve the historic api for the given hash
+		const historicalApi = await api.at(hash);
+
+		/**
+		 * Check two different cases where a historicalApi is needed in order
+		 * to have the correct runtime methods.
+		 *
+		 * a) Does the block use the oldest api where the free balance is found
+		 * using `historicalApi.query.balances.freeBalance`
+		 *
+		 * b) Does the block use an older api where the free balance is within the
+		 * AccountInfo type, but the storage does not yet have the `.at` method.
+		 */
+		if (historicalApi.query.balances.freeBalance) {
+			const [header, free, locks, reserved, nonce] = await Promise.all([
+				api.rpc.chain.getHeader(hash),
+				historicalApi.query.balances.freeBalance(address) as Promise<Balance>,
+				historicalApi.query.balances.locks(address),
+				historicalApi.query.balances.reservedBalance(
+					address
+				) as Promise<Balance>,
+				historicalApi.query.system.accountNonce(address) as Promise<Index>,
+			]);
+
+			// Values dont exist for these historic runtimes
+			const miscFrozen = api.registry.createType('Balance', 0),
+				feeFrozen = api.registry.createType('Balance', 0);
+
+			const at = {
+				hash,
+				height: header.number.toNumber().toString(10),
+			};
+
+			if (locks && free && reserved && nonce) {
+				return {
+					at,
+					nonce,
+					tokenSymbol: token,
+					free,
+					reserved,
+					miscFrozen,
+					feeFrozen,
+					locks,
+				};
+			} else {
+				throw new BadRequest('Account not found');
+			}
+		} else if (
+			!Object.keys(historicalApi.query.system.account).includes('at')
```

Comment:
I assume checking `!historicalApi.query.system.account.at` does not work? if it does i wonder if its slightly faster than pulling all the keys and then doing an O(n) includes

Reply:
It actually does work, thanks for the suggestion. Way cleaner.